### PR TITLE
refactor: leaf_stage range_extension handling

### PR DESCRIPTION
**Pull request by gabriele-0201**
the pr contains the following changes:
+ LeafChanges entries received from the right, if contains an inserted node, will always be used as new base
    Previously, if a new_entry had a bigger new_key, the base was searched through fresh leaves leading to a panic due to a double deletion of the leaf
+ `try_answer_left_neighbor` sends the right neighbor to the left one only if it has finished interacting with the right one
+ a small refactor inside `run_worker` on how the final calls to digest are handled 
+ a link to an issue related to the extension protocol

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/422?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#422** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/422?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#421** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/421?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#420** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/420?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#419** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/419?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/beatree/ops/update/leaf_stage.rs [hunk 1]
```diff
@@ -530,45 +515,45 @@ fn run_worker(
...
+            &mut leaf_updater,
+            has_extended_range,
+            cutoff,
+        );
     }
 
-    // answer any outstanding pending request
+    // Now we are safe to send over our right neighbor to the left one
+    // because we're sure to have finished interacting with it
+    has_finished_with_right_neighbor = true;
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/leaf_stage.rs:548]: this property is more clearly expressed as "has_finished_workload" or something like that, i.e. we've performed all inserts/splits/merges and therefore can safely give over our entire range to the left worker.

File: nomt/src/beatree/ops/update/leaf_stage.rs [hunk 2]
```diff
@@ -308,7 +309,9 @@ fn try_answer_left_neighbor(
 
     let consumed_whole_range = next_separator == worker_params.range.high;
 
-    if !found_next_node && !consumed_whole_range {
+    // wait to respond if no next node was found or if the entire range was consumed but
+    // the worker didn't finish to interact with the right neighbor
+    if !found_next_node && !(consumed_whole_range && has_finished_with_right_neighbor) {
```
Comment (rphmeier) [inline nomt/src/beatree/ops/update/leaf_stage.rs:314]: To me this looks more like `consumed_whole_range` doesn't do what I originally meant it to do. We can probably just remove the notion of `consumed_whole_range` altogether and have `has_finished_with_right_neighbor` do all the heavy lifting, re: https://github.com/thrumdev/nomt/issues/425

The intention is to link up our left and right neighbor once we've finished our work. 

as you've correctly found, there are issues with the current `consumed_whole_range` property.
 1. if we're still interacting with the right neighbor to e.g. complete merges, we can answer too early and cause issues. 
 2. in #425 you note that workers may never consume their whole range (and likely don't)
 
So I suggest a refactor to remove the notion of `consumed_whole_range` and replace it with `has_finished_with_right_neighbor`.

Comment (gabriele-0201) [reply nomt/src/beatree/ops/update/leaf_stage.rs]: `consumed_whole_range` has been removed in #426

**Review by rphmeier — APPROVED** [review]

**Pull request merged**